### PR TITLE
ENH Adds support for no-batch dim in AdaptiveAvgPool1d

### PR DESCRIPTION
**Pull request by thomasjpfan**
Towards https://github.com/pytorch/pytorch/issues/60585

### CODE REVIEW

File: torch/nn/modules/pooling.py [hunk 1]
```diff
@@ -1084,6 +1084,10 @@ class AdaptiveAvgPool1d(_AdaptiveAvgPoolNd):
     Args:
         output_size: the target output size H
 
+    Shape:
+        - Input: :math:`(*, C, L_{in})`, where :math:`*` is 1 dimension or none.
+        - Output: :math:`(*, C, \text{output\_size})`.
```
Comment (jbschlosser) [inline torch/nn/modules/pooling.py:1089]: The docs above reference `H` as the output size, which is confusing. To match the other 1D pooling modules, I suggest using `L_{in}` and `L_{out}`, where `L_{out} = \text{output\_size}`

File: torch/testing/_internal/common_nn.py [hunk 2]
```diff
@@ -1248,6 +1248,13 @@ def fractional_max_pool3d_test(test_case):
             fullname='FractionalMaxPool3d_asymsize')
 
 
+def single_batch_reference_fn(ref_input, parameters, ref_module):
+    """Reference function for module supporting no batch dimensions."""
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:1252]: I like the idea of this function! It could be a little bit clearer that the purpose of this is to run an input through in batched form with a single item (for comparison against true no-batch inputs)

File: aten/src/ATen/native/Pooling.cpp [hunk 3]
```diff
@@ -22,14 +22,14 @@ static void check1d(
 }
 
 Tensor adaptive_avg_pool1d(const Tensor & self, IntArrayRef output_size) {
-  checkDim("adaptive_avg_pool1d", TensorArg(self, "self", 1), 3);
+  checkDimRange("adaptive_avg_pool1d", TensorArg(self, "self", 1), 2, 4);
```
Comment (jbschlosser) [inline aten/src/ATen/native/Pooling.cpp:25]: In other parts of the codebase, it's common to add a comment `/* exclusive */` at the end of this call to `checkDimRange()` to make it clear (see [here](https://github.com/pytorch/pytorch/blob/5da507b57bee31805217889b9f8ff673915dbc3a/aten/src/ATen/native/cudnn/BatchNorm.cpp#L82)).

File: torch/nn/modules/pooling.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nn/modules/pooling.py:1089]: Also wdyt of using `or` instead of `*` here, e.g. `Input: (N, C, L_{in}) or (C, L_{in})`?

**Review by jbschlosser — COMMENTED** [review]
Looking good, couple minor suggestions. Love the `single_batch_reference_fn` idea!

File: torch/nn/modules/pooling.py (hunk 1 above)
Comment (thomasjpfan) [reply torch/nn/modules/pooling.py:1089]: Would the output shape also list out the options?

```
Input: (N, C, L_{in}) or (C, L_{in})
Output: (N, C, L_{out}) or (C, L_{out})
```

I think that looks okay. For the future, if there are only two shape options, list them out?

Comment (jbschlosser) [reply torch/nn/modules/pooling.py:1089]: Yep exactly- there's some precedent for documenting multiple shapes like this (see `nn.Upsample` docs) and I think it's nice and readable

File: torch/nn/modules/adaptive.py [hunk 4]
```diff
@@ -167,6 +167,14 @@ def reset_parameters(self) -> None:
...
 
     def forward(self, input: Tensor, target: Tensor) -> _ASMoutput:
+        have_batch_dim = input.size(0) != self.in_features
+        if not have_batch_dim:
+            if target.dim() != 0:
+                raise RuntimeError('Target must be a scalar when there is no '
+                                   'batch dimension')
+            input = input.unsqueeze(0)
+            target = target.unsqueeze(0)
+
```
Comment (jbschlosser) [inline torch/nn/modules/adaptive.py:177]: Was this change intended for this PR? Looks like this is in `AdaptiveLogSoftmaxWithLoss`

**Comment by jbschlosser** [comment]
test failures look to be spurious :)

File: torch/nn/modules/adaptive.py (hunk 4 above)
Comment (thomasjpfan) [reply torch/nn/modules/adaptive.py]: Yea this was a mistake. (committed to the wrong branch)

File: torch/nn/modules/pooling.py (hunk 1 above)
Comment (thomasjpfan) [reply torch/nn/modules/pooling.py]: Assuming we work out how to deal with `Upsample` + BC breaking, would we list out 3 shapes for no batch and 3 shapes for batches?

**Review by jbschlosser — APPROVED** [review]
LGTM!

File: torch/nn/modules/pooling.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nn/modules/pooling.py]: Ouch, that's a good point- don't think we'd want to do that. For the sake of discussion, let me toss out an idea:
```
([N], C, L), where [N] is an optional batch dimension
```
wdyt about the readability of this? I think this case will be fairly common
